service/history/shard/compatibility.go

```diff
@@ -0,0 +1,283 @@
+						Attributes:    &persistencespb.Predicate_UniversalPredicateAttributes{},
+					},
+				}},
+			}
+		}
+	}
+	return shardInfo
+}
+
+func storeShardInfoCompatibilityCheck(
+	clusterMetadata cluster.Metadata,
+	shardInfo *persistencespb.ShardInfo,
+) *persistencespb.ShardInfo {
+	// TODO this section maintains the forward / backward compatibility
+	//  should be removed once the migration is done
+	//  also see ShardInfoFromBlob
+	allClusterInfo := clusterMetadata.GetAllClusterInfo()
+	shardInfo = storeShardInfoCompatibilityCheckWithoutReplication(shardInfo)
+	shardInfo = storeShardInfoCompatibilityCheckWithReplication(allClusterInfo, shardInfo)
+	return shardInfo
+}
+
+func storeShardInfoCompatibilityCheckWithoutReplication(
+	shardInfo *persistencespb.ShardInfo,
+) *persistencespb.ShardInfo {
+	for queueCategoryID, queueState := range shardInfo.QueueStates {
+		if queueCategoryID == tasks.CategoryIDReplication {
+			continue
+		}
+
+		queueCategory, ok := tasks.GetCategoryByID(queueCategoryID)
+		if !ok {
+			panic(fmt.Sprintf("unable to find queue category by queye category ID: %v", queueCategoryID))
+		}
+
+		// for compatability, update ack level and cluster ack level as well
+		// so after rollback or disabling the feature, we won't load too many tombstones
+		minAckLevel := ConvertFromPersistenceTaskKey(queueState.ExclusiveReaderHighWatermark)
+		for _, readerState := range queueState.ReaderStates {
+			if len(readerState.Scopes) != 0 {
+				minAckLevel = tasks.MinKey(
+					minAckLevel,
+					ConvertFromPersistenceTaskKey(readerState.Scopes[0].Range.InclusiveMin),
+				)
+			}
+		}
+
+		if queueCategory.Type() == tasks.CategoryTypeImmediate && minAckLevel.TaskID > 0 {
+			// for immediate task type, the ack level is inclusive
+			// for scheduled task type, the ack level is exclusive
+			minAckLevel = minAckLevel.Prev()
+		}
+		persistenceAckLevel := convertTaskKeyToPersistenceAckLevel(queueCategory.Type(), minAckLevel)
+
+		shardInfo.QueueAckLevels[queueCategoryID] = &persistencespb.QueueAckLevel{
+			AckLevel:        persistenceAckLevel,
+			ClusterAckLevel: make(map[string]int64),
+		}
+	}
+	return shardInfo
+}
+
+func storeShardInfoCompatibilityCheckWithReplication(
+	allClusterInfo map[string]cluster.ClusterInformation,
+	shardInfo *persistencespb.ShardInfo,
+) *persistencespb.ShardInfo {
+	shardInfo = trimShardInfo(allClusterInfo, shardInfo)
+	if shardInfo.QueueStates == nil {
+		return shardInfo
+	}
+	queueStates, ok := shardInfo.QueueStates[tasks.CategoryIDReplication]
+	if !ok {
+		return shardInfo
+	}
+
+	for readerID, readerState := range queueStates.ReaderStates {
+		clusterID, _ := ReplicationReaderIDToClusterShardID(readerID)
+		clusterName, _, _ := ClusterNameInfoFromClusterID(allClusterInfo, clusterID)
+		queueAckLevel, ok := shardInfo.QueueAckLevels[tasks.CategoryIDReplication]
+		if !ok {
+			queueAckLevel = &persistencespb.QueueAckLevel{
+				AckLevel:        0,
```

Comment:
just double checking: replication doesn't use this?

Reply:
i do not believe replication use ack level var (this should be used by transfer Q, time Q, etc)
replication acks are per cluster, so only `ClusterAckLevel map[string]int64` should be used